animated-transformer/src/lib/transformer/gpt2.ts

```diff
@@ -0,0 +1,574 @@
+        model.config.spec.computeSpec.layerNormEpsilon);
+    let finalOutput = lastBlock;
+    finalOutput.seqOutput = finalSeqOutput;
+    finalOutput.seqInput = lastBlock.seqOutput;
+    compute.layers.push(finalOutput);
+  }
+  
+  return compute;
+}
+
+/** Batch compute the loss for the last token of a transformer.
+ *
+ * params: transformer parameters.
+ * tokenEmb: embeddings for all tokens.
+ * targetTokenIdxs: a one-hot token vector for the correct token.
+ */
+export function lastTokenLogits(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation
+): GTensor<'batch' | 'tokenId'> {
+  const lastLayer = computation.layers[computation.layers.length - 1];
+  const positionParams = lastLayer.seqOutput.unstack('pos');
+  const lastPosParams = positionParams[positionParams.length - 1];
+  const logits = lastPosParams.contract(model.params.tokenEmbedding, ['inputRep']);
+  return logits;
+}
+
+/**
+ * Returns the average per example loss for the last token prediction.
+ */
+export function lastTokenCrossEntropyLoss(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation,
+  targetTokenIdxs: GTensor<'batch'>
+): tf.Scalar {
+  const logits = lastTokenLogits(model, computation);
+  const logProbs = logits.softmax('tokenId').log();
+  const nTokens = model.params.tokenEmbedding.dim.tokenId.size;
+  const oneHotToken = new GTensor(oneHot(targetTokenIdxs.tensor, nTokens), ['batch', 'tokenId']);
+  const crossEntopy = logProbs.pointwiseMul(oneHotToken);
+  return (
+    crossEntopy
+      .sumOverDims(['batch', 'tokenId'])
+      // ._tfScalarMul(tf.scalar(-1))
+      ._tfScalarDiv(tf.scalar(targetTokenIdxs.dim.batch.size * -1)).tensor as tf.Scalar
+  );
+  // const squaredError = signedDelta.pointwiseMul(signedDelta);
+  // const loss = squaredError.sumOverDims(['batch', 'token']);
+  // return loss.tensor;
+}
+
+/** Batch compute the top prediction from the last token of a transformer.
+ *
+ * params: transformer parameters.
+ * tokenEmb: embeddings for all tokens.
+ * targetTokenIdxs: a one-hot token vector for the correct token.
+ */
+export function transformerTopPrediction(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation
+): GTensor<'batch'> {
+  const dotProd = lastTokenLogits(model, computation);
+  return dotProd.argMax('tokenId');
+}
+
+export function transformerAccuracy(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation,
+  targetTokenIdxs: GTensor<'batch'>
+): tf.Scalar {
+  const predictions = transformerTopPrediction(model, computation);
+  return predictions
+    .pointwiseEqual(targetTokenIdxs)
+    .sumOverDims(['batch'])
+    .tensor.div(tf.scalar(targetTokenIdxs.dim.batch.size));
+}
+
+export function addPosEmbeddings (
```

Comment:
indentation is odd within this function def.

Reply:
Fixed it!